x-pack/plugin/ml-feature-index-builder/src/main/java/org/elasticsearch/xpack/ml/featureindexbuilder/support/JobValidator.java

```diff
@@ -0,0 +1,194 @@
+                aggregationSourceFieldNames.put(valueSourceAggregation.getName(), valueSourceAggregation.field());
+                aggregationTypes.put(valueSourceAggregation.getName(), valueSourceAggregation.getType());
+            } else {
+                // execution should not reach this point
+                listener.onFailure(new RuntimeException("Unsupported aggregation type [" + agg.getType() + "]"));
+                return;
+            }
+        }
+
+        Map<String, String> allFieldNames = new HashMap<>();
+        allFieldNames.putAll(aggregationSourceFieldNames);
+        allFieldNames.putAll(fieldNamesForGrouping);
+
+        getSourceFieldMappings(config.getIndexPattern(), allFieldNames.values().toArray(new String[allFieldNames.size()]),
+                ActionListener.wrap(sourceMappings -> {
+                    Map<String, String> targetMapping = new HashMap<>();
+
+                    aggregationTypes.forEach((targetFieldName, aggregationName) -> {
+                        String sourceFieldName = aggregationSourceFieldNames.get(targetFieldName);
+                        String destinationMapping = Aggregations.resolveTargetMapping(aggregationName, sourceMappings.get(sourceFieldName));
+
+                        logger.debug("[" + config.getId() + "] Deduced mapping for: [" + targetFieldName + "], agg type [" + aggregationName
+                                + "] to [" + destinationMapping + "]");
+                        if (destinationMapping != null) {
+                            targetMapping.put(targetFieldName, destinationMapping);
+                        } else {
+                            logger.warn("[" + config.getId() + "] Failed to deduce mapping for [" + targetFieldName
+                                    + "], fall back to double.");
+                            targetMapping.put(targetFieldName, "double");
+                        }
+                    });
+
+                    fieldNamesForGrouping.forEach((targetFieldName, sourceFieldName) -> {
+                        String destinationMapping = sourceMappings.get(sourceFieldName);
+                        logger.debug(
+                                "[" + config.getId() + "] Deduced mapping for: [" + targetFieldName + "] to [" + destinationMapping + "]");
+                        if (destinationMapping != null) {
+                            targetMapping.put(targetFieldName, destinationMapping);
+                        } else {
+                            logger.warn("[" + config.getId() + "] Failed to deduce mapping for [" + targetFieldName
+                                    + "], fall back to keyword.");
+                            targetMapping.put(targetFieldName, "keyword");
+                        }
+                    });
+
+                    listener.onResponse(targetMapping);
+                }, e2 -> {
+                    listener.onFailure(e2);
+                }));
```

Comment:
#nit 
`e2` implies that there is an `e1` somewhere, but I cannot see it.

Reply:
;-) (e1 was) lost during re-factoring